// mobx-react-form/lib/shared/Utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _trimEnd2 = require('lodash/trimEnd');

var _trimEnd3 = _interopRequireDefault(_trimEnd2);

var _each2 = require('lodash/each');

var _each3 = _interopRequireDefault(_each2);

var _isNil2 = require('lodash/isNil');

var _isNil3 = _interopRequireDefault(_isNil2);

var _head2 = require('lodash/head');

var _head3 = _interopRequireDefault(_head2);

var _split2 = require('lodash/split');

var _split3 = _interopRequireDefault(_split2);

var _merge3 = require('lodash/merge');

var _merge4 = _interopRequireDefault(_merge3);

var _omit2 = require('lodash/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _has2 = require('lodash/has');

var _has3 = _interopRequireDefault(_has2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _mobx = require('mobx');

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _parser = require('../parser');

var _parser2 = _interopRequireDefault(_parser);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
  Field Utils
*/
exports.default = {
  on: function on(event, callback) {
    var _this = this;

    var path = (0, _has3.default)(this, 'isField') ? this.path : true;
    return (0, _mobx.observe)(this.state.events.$running, function (change) {
      return event === change.name && change.newValue !== false && _this.state.events.$running[event] === path && callback({
        path: _this.state.events.$running[event],
        change: (0, _omit3.default)(change, 'type', 'object'),
        form: _this,
        event: event
      });
    });
  },


  /**
   Observer
   */
  observe: function observe(_ref) {
    var _ref$path = _ref.path,
        path = _ref$path === undefined ? null : _ref$path,
        key = _ref.key,
        call = _ref.call;

    var $field = (0, _has3.default)(this, 'isField') ? this : this.select(path);
    var params = { form: this.state.form, path: $field.path, $field: $field };
    var disposer = key + '@' + $field.path;

    (0, _merge4.default)(this.state.disposers, _defineProperty({}, disposer, key === 'fields' ? $field.fields.observe(function (change) {
      return call.apply(null, [_extends({}, params, { change: change })]);
    }) : (0, _mobx.observe)($field, key, function (change) {
      return call.apply(null, [_extends({}, params, { change: change })]);
    })));
  },


  /**
   Disposer
   */
  dispose: function dispose(key) {
    var path = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

    var $path = _parser2.default.parsePath(path || this.path);
    this.state.disposers[key + '@' + $path]();
    delete this.state.disposers[key + '@' + $path];
  },


  /**
   Fields Selector
   */
  select: function select(path) {
    var fields = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
    var isStrict = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

    var $path = _parser2.default.parsePath(path);

    var keys = (0, _split3.default)($path, '.');
    var head = (0, _head3.default)(keys);

    keys.shift();

    var $fields = (0, _isNil3.default)(fields) ? this.fields.get(head) : fields.get(head);

    var stop = false;
    (0, _each3.default)(keys, function ($key) {
      if (stop) return;
      if ((0, _isNil3.default)($fields)) {
        $fields = undefined;
        stop = true;
      } else {
        $fields = $fields.fields.get($key);
      }
    });

    if (isStrict) _utils2.default.throwError(path, $fields);

    return $fields;
  },


  /**
    Get Container
   */
  container: function container() {
    var path = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    var $path = path || this.path || '';
    var cpath = (0, _trimEnd3.default)($path.replace(new RegExp('/[^./]+$/'), ''), '.');
    return this.select(cpath, null, false);
  },


  /**
   * Iterates deeply over fields and invokes `iteratee` for each element.
   * The iteratee is invoked with three arguments: (value, index|key, depth).
   *
   * @param {Function} iteratee The function invoked per iteration.
   * @param {Array|Object} [fields=form.fields] fields to iterate over.
   * @param {number} [depth=1] The recursion depth for internal use.
   * @returns {Array} Returns [fields.values()] of input [fields] parameter.
   * @example
   *
   * JSON.stringify(form)
   * // => {
     *   "fields": {
     *     "state": {
     *       "fields": {
     *         "city": {
     *           "fields": { "places": {
     *                "fields": {},
     *                "key": "places", "path": "state.city.places", "$value": "NY Places"
     *              }
     *           },
     *           "key": "city", "path": "state.city", "$value": "New York"
     *         }
     *       },
     *       "key": "state", "path": "state", "$value": "USA"
     *     }
     *   }
     * }
   *
   * const data = {};
   * form.forEach(field => data[field.path] = field.value);
   * // => {
     *   "state": "USA",
     *   "state.city": "New York",
     *   "state.city.places": "NY Places"
     * }
   *
   */
  forEach: function forEach(iteratee) {
    var _this2 = this;

    var fields = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
    var depth = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

    var $fields = fields || this.fields;
    (0, _each3.default)($fields.values(), function (field, index) {
      iteratee(field, index, depth);

      if (field.fields.size !== 0) {
        _this2.forEach(iteratee, field.fields, depth + 1);
      }
    });
  }
};
module.exports = exports['default'];
