// mobx-react-form/lib/Base.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _desc, _value, _class;

var _mobx = require('mobx');

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Base = (_class = function () {
  function Base() {
    var _this = this;

    _classCallCheck(this, Base);

    this.onClear = function (e) {
      e.preventDefault();
      _this.clear(true);
    };

    this.onReset = function (e) {
      e.preventDefault();
      _this.reset(true);
    };

    this.onAdd = function (e) {
      var val = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

      e.preventDefault();
      _this.add(_utils2.default.isEvent(val) ? null : val);
    };

    this.onDel = function (e) {
      var path = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

      e.preventDefault();
      _this.del(_utils2.default.isEvent(path) ? _this.path : path);
    };
  }

  _createClass(Base, [{
    key: 'hasIncrementalNestedFields',
    get: function get() {
      return _utils2.default.hasIntKeys(this.fields) && this.fields.size;
    }
  }, {
    key: 'hasNestedFields',
    get: function get() {
      return this.fields.size !== 0;
    }

    /**
      Event: On Clear
    */


    /**
      Event: On Reset
    */


    /**
      Event: On Add
    */


    /**
      Event: On Del
    */

  }]);

  return Base;
}(), (_applyDecoratedDescriptor(_class.prototype, 'hasIncrementalNestedFields', [_mobx.computed], Object.getOwnPropertyDescriptor(_class.prototype, 'hasIncrementalNestedFields'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'hasNestedFields', [_mobx.computed], Object.getOwnPropertyDescriptor(_class.prototype, 'hasNestedFields'), _class.prototype)), _class);
exports.default = Base;
module.exports = exports['default'];
